src/main.rs

```diff
@@ -205,16 +205,40 @@ impl App {
             let document_links = self.document_links(&params).unwrap_or_default();
             self.reply(Response::new_ok(id, document_links));
         } else if let Some((id, params)) = cast::<Formatting>(&mut req) {
-            let changes = if let Some((ast, code, _)) = self.files.get(&params.text_document.uri) {
-                let fmt = nixpkgs_fmt::reformat_node(&ast.node());
-                vec![TextEdit {
-                    range: utils::range(&code, TextRange::up_to(ast.node().text().len())),
-                    new_text: fmt.text().to_string(),
-                }]
+            if let Some((ast, code, _)) = self.files.get(&params.text_document.uri) {
+                let node = ast.node();
+                let range = utils::range(&code, TextRange::up_to(node.text().len()));
+                if cfg!(feature = "alejandra") {
+                    #[cfg(feature = "alejandra")]
+                    match alejandra::format::in_memory(
+                        params.text_document.uri.to_string(),
+                        code.to_string(),
+                    ) {
+                        (alejandra::format::Status::Changed(true), new_text) => {
+                            self.reply(Response::new_ok(id, vec![TextEdit { range, new_text }]))
```

Comment:
We're implementing "partial" edits in #87 for nixpkgs-fmt, i.e. only the actual diffs, but not the reformatted file is returned.
AFAIU this also returns the entire file.

If we add a new formatter we should do it properly from the beginning, i.e. only return diffs. Otherwise it might be painful in the editor if e.g. `all-packages.nix` is attempted to be reformatted.

Reply:
@rvolosatovs  `alejandra::format::in_memory` is all we have

I understand why sending small `TextEdit`s instead of a `TextEdit` with the whole file can be a good thing. The most simple approach I can think of is:
- Use `alejandra::format::in_memory` to get a string with the formatted text
- Use a diffing library to see which lines are equal in the input text and in the formatted text
- Emit `TextEdit`s for the ranges that are different

I'm not sure if that is good enough for this use case, I'm not familiar with LSP

> would you be open to having that feature implemented in the first place?

That feature is more related to this repository than to Alejandra, so it would be better to have it here and let Alejandra just focus on being good at formatting

Reply:
Not a maintainer of this anymore, so I don't have a strong opinion. Please discuss that with the current maintainers :)